### PR TITLE
Infiniband support for VMs

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

File: lxd/device/infiniband_physical.go [hunk 1]
```diff
@@ -4,10 +4,12 @@ import (
 	"fmt"
 
 	deviceConfig "github.com/lxc/lxd/lxd/device/config"
+	pcidev "github.com/lxc/lxd/lxd/device/pci"
```
Comment (tomponline) [inline lxd/device/infiniband_physical.go:7]: can we import this as just `pci`?

File: lxd/device/infiniband_sriov.go [hunk 2]
```diff
@@ -4,11 +4,13 @@ import (
 	"fmt"
 
 	deviceConfig "github.com/lxc/lxd/lxd/device/config"
+	pcidev "github.com/lxc/lxd/lxd/device/pci"
```
Comment (tomponline) [inline lxd/device/infiniband_sriov.go:7]: Can we import this as just `pci`?

File: lxd/device/infiniband_physical.go (hunk 1 above)
Comment (tomponline) [reply lxd/device/infiniband_physical.go:7]: Oh I see you're just being consistent with other imports, makes sense :)

File: lxd/device/infiniband_sriov.go (hunk 2 above)
Comment (tomponline) [reply lxd/device/infiniband_sriov.go:7]: Oh I see you're just being consistent with other imports, makes sense :)

**Review by tomponline — APPROVED** [review]

**Comment by monstermunchkin** [comment]
The current status is that it's not working.

For SR-IOV I'm getting:
```
ubuntu@lantea:~$ lxc config device add v1 ib infiniband parent=ibp4s0 nictype=sriov
Device ib added to v1
ubuntu@lantea:~$ lxc start v1
Error: Failed to start device "ib": Failed growing available VFs from 0 to 31 on device "ibp4s0": write /sys/class/net/ibp4s0/device/sriov_numvfs: no such file or directory
Try `lxc info --show-log v1` for more info
```

That doesn't make sense to me since the file _does_ exist.

For phyiscal I'm getting:
```
ubuntu@lantea:~$ lxc config device add v1 ib infiniband parent=ibp4s0 nictype=physical
Device ib added to v1
ubuntu@lantea:~$ lxc start v1
Error: Failed setting up device via monitor: Failed setting up device "ib": Failed adding NIC device: vfio 0000:04:00.0: Failed to set up TRIGGER eventfd signaling for interrupt INTX-0: VFIO_DEVICE_SET_IRQS failure: Device or resource busy
Try `lxc info --show-log v1` for more info
```

**Comment by tomponline** [comment]
Have you confirmed that the container functionality is working first, so we have a line in the sand that we know SRIOV and physical passthrough is working for containers?

**Comment by monstermunchkin** [comment]
> Have you confirmed that the container functionality is working first, so we have a line in the sand that we know SRIOV and physical passthrough is working for containers?

Physical works on containers. When trying SR-IOV, I get:
```
ubuntu@lantea:~$ lxc start c1
Error: Failed to start device "ib": All virtual functions on parent device are already in use
Try `lxc info --show-log c1` for more info
```

**Comment by tomponline** [comment]
> > Have you confirmed that the container functionality is working first, so we have a line in the sand that we know SRIOV and physical passthrough is working for containers?
> 
> Physical works on containers. When trying SR-IOV, I get:
> 
> ```
> ubuntu@lantea:~$ lxc start c1
> Error: Failed to start device "ib": All virtual functions on parent device are already in use
> Try `lxc info --show-log c1` for more info
> ```

What does `ip l` show?

**Comment by monstermunchkin** [comment]
> What does `ip l` show?

```
ubuntu@lantea:~$ ip  l
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN mode DEFAULT group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: enp2s0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc fq_codel state DOWN mode DEFAULT group default qlen 1000
    link/ether 00:8e:25:70:00:48 brd ff:ff:ff:ff:ff:ff
3: enp9s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel master br0 state UP mode DEFAULT group default qlen 1000
    link/ether d0:50:99:c2:92:63 brd ff:ff:ff:ff:ff:ff
4: enp5s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether 80:61:5f:03:9a:85 brd ff:ff:ff:ff:ff:ff
5: enp13s0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP mode DEFAULT group default qlen 1000
    link/ether d0:50:99:c2:92:64 brd ff:ff:ff:ff:ff:ff
6: enp129s0f0np0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether b8:ce:f6:09:fa:2c brd ff:ff:ff:ff:ff:ff
7: enp129s0f1np1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP mode DEFAULT group default qlen 1000
    link/ether b8:ce:f6:09:fa:2d brd ff:ff:ff:ff:ff:ff
9: br0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default qlen 1000
    link/ether d0:50:99:c2:92:63 brd ff:ff:ff:ff:ff:ff
33: lxdbr0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether 00:16:3e:c7:d4:de brd ff:ff:ff:ff:ff:ff
36: enp4s0d1: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN mode DEFAULT group default qlen 1000
    link/ether 00:02:c9:a0:00:91 brd ff:ff:ff:ff:ff:ff
37: ibp4s0: <BROADCAST,MULTICAST> mtu 4092 qdisc noop state DOWN mode DEFAULT group default qlen 256
    link/infiniband 80:00:02:30:fe:80:00:00:00:00:00:00:00:02:c9:03:00:a0:00:91 brd 00:ff:ff:ff:ff:12:40:1b:ff:ff:00:00:00:00:00:00:ff:ff:ff:ff
```

**Comment by tomponline** [comment]
OK so SRIOV isn't enable on that physical NIC yet, as it doesn't show any VFs.

**Comment by tomponline** [comment]
@monstermunchkin any luck?

**Comment by monstermunchkin** [comment]
So, the current state is that SR-IOV now works for VMs. However, I'm still getting the following error when doing `nictype=physical`:

```
ubuntu@lantea:~$ lxc config device add v1 ib infiniband parent=ibp4s0 nictype=physical
Device ib added to v1
ubuntu@lantea:~$ lxc start v1
Error: Failed setting up device via monitor: Failed setting up device "ib": Failed adding NIC device: vfio 0000:04:00.0: Failed to set up TRIGGER eventfd signaling for interrupt INTX-0: VFIO_DEVICE_SET_IRQS failure: Device or resource busy
Try `lxc info --show-log v1` for more info
```

This also happens when using the `pci` device type and passing `address=0000:04:00.0`. I haven't found a solution for this yet. @tomponline @stgraber have you come across this error before?

**Comment by tomponline** [comment]
@monstermunchkin has the device been unbound from the host before passing in?

**Comment by monstermunchkin** [comment]
> @monstermunchkin has the device been unbound from the host before passing in?

Yes, both `pci` and `infiniband` device types unbind the device.

**Comment by stgraber** [comment]
@monstermunchkin when testing passthrough, did you disable sr-iov?

I don't think you'll be able to move the PF to a VM when it has VFs.

**Comment by monstermunchkin** [comment]
> when testing passthrough, did you disable sr-iov?

I believe I did as I thought that was causing the issue. I'll check again once lantea is available.

**Comment by monstermunchkin** [comment]
@stgraber I just checked again, and sr-iov is disabled. The error remains the same:
```
qemu-system-x86_64:/var/snap/lxd/common/lxd/logs/v1/qemu.conf:265: vfio 0000:04:00.0: Failed to set up TRIGGER eventfd signaling for interrupt INTX-0: VFIO_DEVICE_SET_IRQS failure: Device or resource busy
```

**Comment by stgraber** [comment]
@monstermunchkin is this branch up to date?

**Comment by stgraber** [comment]
@monstermunchkin with your current branch, using plain sriov which you said works fine, I'm getting:
```
root@lantea:~# lxc start u1
root@lantea:~# lxc start v1
Error: Failed to start device "ib0": no matching virtual function found
Try `lxc info --show-log v1` for more info
```

So works fine with the existing container code, does not work with the VM.

So I'm guessing this branch isn't up to date and is missing changes. I'll wait for you to update it and rebase it as well as re-validate it on SR-IOV before I dig into what's going on with physical.

In general, the test for this should be:
```
rmmod mlx4_ib
rmmod mlx4_en
rmmod mlx4_core
modprobe mlx4_core num_vfs=16 probe_vf=16
lxc init images:ubuntu/22.04/cloud v1 --vm
lxc config device add v1 ib0 infiniband nictype=sriov parent=ibp4s0
lxc start v1
lxc init images:ubuntu/22.04 u1
lxc config device add u1 ib0 infiniband nictype=sriov parent=ibp4s0
lxc start u1
```

**Comment by monstermunchkin** [comment]
> In general, the test for this should be:
> 
> ```
> rmmod mlx4_ib
> rmmod mlx4_en
> rmmod mlx4_core
> modprobe mlx4_core num_vfs=16 probe_vf=16
> ```

@stgraber I had to drop `probe_vf` the last time I tested it. Otherwise it would return an error stating that all VFs are already in use.

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by monstermunchkin** [comment]
@stgraber did you mange to get `nictype=physical` working?

**Pull request merged**